File: lang/csharp/src/vpkpp/Format/GMA.cs
using System;

namespace sourcepp.vpkpp.Format;

using OpenPropertyRequest = Func<PackFile, OpenProperty, byte[]>;

using EntryCallback = Action<string, Entry>;

public class GMA : PackFile
{
	protected GMA(nint handle, bool managed = true) : base(handle, managed)
	{
	}

	public new static GMA? Open(string path, EntryCallback? callback = null, OpenPropertyRequest? _ = null)
	{
		var handle = DLL.vpkpp_gma_open(path, callback is not null ? (entryPath, entry) =>
		{
			callback(entryPath, new Entry(entry, false));
		} : null);
		return handle == nint.Zero ? null : new GMA(handle);
	}
}

File: lang/c/src/vpkppc/format/GMA.cpp
#include <vpkppc/format/GMA.h>

#include <vpkpp/format/GMA.h>

#include <sourceppc/Helpers.h>
#include <vpkppc/Convert.hpp>

using namespace sourceppc;
using namespace vpkpp;

SOURCEPP_API vpkpp_pack_file_handle_t vpkpp_gma_open(const char* path, vpkpp_entry_callback_t callback) {
	SOURCEPP_EARLY_RETURN_VAL(path, nullptr);

	auto packFile = GMA::open(path, callback ? [callback](const std::string& path, const Entry& entry) {
		callback(path.c_str(), const_cast<Entry*>(&entry));
	} : static_cast<PackFile::EntryCallback>(nullptr));
	if (!packFile) {
		return nullptr;
	}
	return packFile.release();
}
